src/main/java/tech/ydb/io/r2dbc/executor/QueryExecutor.java

```diff
@@ -0,0 +1,261 @@
+
+    public QueryExecutor(YdbContext ydbContext, YdbConnectionState ydbConnectionState) {
+        this.ydbConnectionState = ydbConnectionState;
+        this.statementTimeout = ydbContext.getDefaultTimeout();
+    }
+
+    public Flux<YdbResult> executeDataQuery(String yql, Params params, List<OperationType> operationTypes) {
+        return fluxWithState(connectionState -> connectionState.getSession()
+                .flatMapMany(session ->
+                        Mono.fromFuture(session.executeDataQuery(yql, connectionState.txControl(), params,
+                                        withStatementTimeout(new ExecuteDataQuerySettings())))
+                                .flatMapMany(dataQueryResult -> {
+                                    Mono<DataQueryResult> dataQueryResultMono =
+                                            ResultExtractor.extract(dataQueryResult);
+
+                                    return dataQueryResultMono.flatMapMany(result -> {
+                                                List<YdbResult> results = new ArrayList<>();
+                                                for (int opIndex = 0, resSetIndex = 0; opIndex < operationTypes.size(); opIndex++) {
+                                                    results.add(switch (operationTypes.get(opIndex)) {
+                                                        case SELECT -> new YdbResult(result.getResultSet(resSetIndex++));
+                                                        case UPDATE -> YdbResult.UPDATE_RESULT;
+                                                        case SCHEME -> throw new IllegalStateException("DDL operation" +
+                                                                " not" +
+                                                                " support in" +
+                                                                " " +
+                                                                "executeDataQuery");
+                                                    });
+                                                }
+
+                                                return Flux.fromIterable(results);
+                                            }).doOnComplete(() -> {
+                                                YdbConnectionState nextState = connectionState.withDataQuery(
+                                                        dataQueryResult.getValue().getTxId(),
+                                                        session
+                                                );
+                                                if (!nextState.isInTransaction()) {
+                                                    session.close();
+                                                }
+                                                updateState(nextState);
+                                            }).doOnError((unused) -> session.close())
+                                            .doOnCancel(session::close);
```

Comment:
see pg

Reply:
Change to DiscardOnCancelSubscriber like in pg-r2dbc